pkg/app_watcher/app_watcher.go

```diff
@@ -0,0 +1,154 @@
+	if !canProcessApp(obj) {
+		return
+	}
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		log.Error().Err(err).Msg("appwatcher: could not get key for added application")
+	}
+	log.Debug().Str("key", key).Msg("appwatcher: onApplicationAdded")
+	ctrl.cfg.VcsToArgoMap.AddApp(obj.(*appv1alpha1.Application))
+}
+
+func (ctrl *ApplicationWatcher) onApplicationUpdated(old, new interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(new)
+	if err != nil {
+		log.Warn().Err(err).Msg("appwatcher: could not get key for updated application")
+	}
+	oldApp := old.(*appv1alpha1.Application)
+	newApp := new.(*appv1alpha1.Application)
+
+	// We want to update when any of Source or Sources parameters has changed
+	if !reflect.DeepEqual(oldApp.Spec.Source, newApp.Spec.Source) || !reflect.DeepEqual(oldApp.Spec.Sources, newApp.Spec.Sources) {
+		log.Debug().Str("key", key).Msg("appwatcher: onApplicationUpdated")
+		ctrl.cfg.VcsToArgoMap.UpdateApp(old.(*appv1alpha1.Application), new.(*appv1alpha1.Application))
+	}
+
+}
+
+func (ctrl *ApplicationWatcher) onApplicationDeleted(obj interface{}) {
+	if !canProcessApp(obj) {
+		return
+	}
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		log.Warn().Err(err).Msg("appwatcher: could not get key for deleted application")
+	}
+
+	log.Debug().Str("key", key).Msg("appwatcher: onApplicationDeleted")
+	ctrl.cfg.VcsToArgoMap.DeleteApp(obj.(*appv1alpha1.Application))
+}
+
+/*
+This Go function, named newApplicationInformerAndLister, is part of the ApplicationWatcher struct. It sets up a Kubernetes SharedIndexInformer and a Lister for Argo CD Applications.
+A SharedIndexInformer is used to watch changes to a specific type of Kubernetes resource in an efficient manner. It significantly reduces the load on the Kubernetes API server by sharing and caching watches between all controllers that need to observe the object.
+Listers use the data from the informer's cache to provide a read-optimized view of the cache which reduces the load on the API Server and hides some complexity.
+*/
+func (ctrl *ApplicationWatcher) newApplicationInformerAndLister(refreshTimeout time.Duration) (cache.SharedIndexInformer, applisters.ApplicationLister) {
+	informer := informers.NewApplicationInformer(ctrl.applicationClientset, "", refreshTimeout,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+	)
+
+	lister := applisters.NewApplicationLister(informer.GetIndexer())
+	informer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    ctrl.onApplicationAdded,
+			UpdateFunc: ctrl.onApplicationUpdated,
+			DeleteFunc: ctrl.onApplicationDeleted,
+		},
+	)
+	return informer, lister
+}
+
+func canProcessApp(obj interface{}) bool {
+	app, ok := obj.(*appv1alpha1.Application)
+	if !ok {
+		return false
+	}
+
+	for _, src := range app.Spec.Sources {
+		if isGitRepo(src.RepoURL) {
+			return true
+		}
+	}
+
+	if !isGitRepo(app.Spec.Source.RepoURL) {
```

Comment:
might want to check `app.Spec.Source != nil`. not sure if it would happen in real life, but why worry?